client/allocrunner/taskrunner/logmon_hook.go

```diff
@@ -94,21 +98,29 @@ func reattachConfigFromHookData(data map[string]string) (*plugin.ReattachConfig,
 
 func (h *logmonHook) Prestart(ctx context.Context,
 	req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	var reattachConfig *plugin.ReattachConfig
+	var err error
 
-	reattachConfig, err := reattachConfigFromHookData(req.PreviousState)
-	if err != nil {
-		h.logger.Error("failed to load reattach config", "error", err)
-		return err
-	}
+	// If the task was restarted and the logmon process is still running logmon
+	// needs the start RPC called again to open the fifo
+	if !h.exited || h.logmonPluginClient == nil || h.logmonPluginClient.Exited() {
```

Comment:
Are you sure this is the right fix? I thought `launchLogMon` was intended to be idempotent (if we're already attached: do nothing. reattach if possible, if not: launch). I'd much rather have an idempotent API than track multiple bits of state across hook methods as hooks are difficult to fully integration test (it's hard to test all the state permutations with realistic dependencies and external process management).

It's much easier to test an idempotent API as well: no matter how many times I call this thing a logmon process should exist.

If that's not possible, this logic appears correct, but I'll be honest: it took me a few readings to fully understand the exact conditions. Launching something only when it *hasn't* exited is strange logic at a glance. Renaming `exited -> taskExited` might make it more obvious.

Reply:
There are three ways (that I know of) that Prestart can be called. When starting a new task, when recovering an existing task and when restarting a task. In each case logmon needs to do something subtly different. In the first case we launch logmon and call start, in the second we reattach and return and in the third case which I’m modeling here we just need to call start.

Making this restarted flag part of the PrestartHookRequest would actually be really nice imo. Or possibly an enum on what caused the hook?

Reply:
> Making this restarted flag part of the PrestartHookRequest would actually be really nice imo. Or possibly an enum on what caused the hook?

That's an interesting idea, but exposes so much more complexity to hooks. If it's necessary or decreases hook code complexity we can do it, but the hope was that hooks would try to be idempotent and only use state as an optimization (eg skip re-downloading artifacts if they're marked as done).

Reply:
Hm, "recovering an existing task" in particular isn't a useful thing to expose to hooks as you might be recovering but running this prestart hook for the first time (eg nomad restarts before all prestart hooks complete on first run). So (to make up enum values) `Recovering` could be `FirstRun` or `Restarting` depending on when the agent shutdown occurred.

A `FirstRun` vs `Restarting` flag might be useful, but `PrestartDone` already covers the common "run once" case. So I don't know how often it would get used. Also since the Nomad agent could die mid-FirstRun, FirstRun logic always needs to be idempotent.

Reply:
Maybe the better fix here is to make logmon's Start RPC idempotent. The we can drop the restarting case and simply always call it, even during reattachment.